### PR TITLE
Error when value is an object - validating api data

### PR DESCRIPTION
**Pull request by daycry**
For example:
```json
{
	"configuration" :
        {
	        "consumerKey1" : "aa",
	        "consumerSecret" : "bb",
	        "oauthToken" : "cc",
	        "oauthTokenSecret" : "dd"
        }
}
```

**Description**
Explain what you have changed, and why.

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by samsonasik** [comment]
Please provide unit test and run cs fix:

```
vendor/bin/php-cs-fixer fix system/Validation/Validation.php
```

**Comment by daycry** [comment]
Done.

**Comment by samsonasik** [comment]
please run cs fix

**Comment by daycry** [comment]
Done, sorry.

**Comment by samsonasik** [comment]
cs still error, see https://github.com/codeigniter4/CodeIgniter4/pull/5142/checks?check_run_id=3733816746#step:8:20

**Comment by daycry** [comment]
ok,
I have launched the processes for fix it

- vendor/bin/php-cs-fixer fix .\tests\system\Validation\ValidationTest.php
- vendor/bin/php-cs-fixer fix .\tests\_support\Validation\TestRules.php

File: tests/system/Validation/ValidationTest.php [hunk 1]
```diff
@@ -628,6 +628,33 @@ public function testTagReplacement()
         $this->assertSame($expected, $errors['Username']);
     }
 
+    public function testRulesForObjectField()
+    {
+        $config          = new App();
+        $config->baseURL = 'http://example.com/';
```
Comment (kenjis) [inline tests/system/Validation/ValidationTest.php:634]: Why do you need `$config`?
It seems it is not used.

File: system/Validation/Validation.php [hunk 2]
```diff
@@ -291,6 +291,8 @@ protected function processRules(string $field, ?string $label, $value, $rules =
                 // if the $value is an array, convert it to as string representation
                 if (is_array($value)) {
                     $value = '[' . implode(', ', $value) . ']';
+                } elseif (\is_object($value)) {
+                    $value = \json_encode($value);
```
Comment (paulbalandan) [inline system/Validation/Validation.php:295]:
```suggestion
                } elseif (is_object($value)) {
                    $value = json_encode($value);
```

File: tests/system/Validation/ValidationTest.php [hunk 3]
```diff
@@ -628,6 +628,33 @@ public function testTagReplacement()
...
+
+        $this->validation->setRules([
+            'configuration' => 'required|check_object_rule',
+        ]);
+
+        $data = (object) ['configuration' => (object) ['first' => 1, 'second' => 2]];
+        $this->validation->run((array) $data);
```
Comment (paulbalandan) [inline tests/system/Validation/ValidationTest.php:641]: Why do you need to cast this first to object then recast to array? Casting is not recursive. https://3v4l.org/gMad5
```suggestion
        $data = ['configuration' => (object) ['first' => 1, 'second' => 2]];
        $this->validation->run($data);
```

File: tests/system/Validation/ValidationTest.php [hunk 4]
```diff
@@ -628,6 +628,33 @@ public function testTagReplacement()
...
+        $this->assertSame([], $this->validation->getErrors());
+
+        $this->validation->reset();
+
+        $this->validation->setRules([
+            'configuration' => 'required|check_object_rule',
+        ]);
+
+        $data = (object) ['configuration' => (object) ['first1' => 1, 'second' => 2]];
+        $this->validation->run((array) $data);
```
Comment (paulbalandan) [inline tests/system/Validation/ValidationTest.php:651]: Same comment as above

File: tests/system/Validation/ValidationTest.php (hunk 3 above)
Comment (daycry) [reply tests/system/Validation/ValidationTest.php:641]: If you look at the method, it only accepts arrays.
I Know that isn't recursive. Configuration attribute is a Object.

Comment (daycry) [reply tests/system/Validation/ValidationTest.php:638]: The change that I propose can be used to validate that what an object has are the mandatory fields, for example:

_**public function check_token_configuration( Object $value, string $socialNetwork, array $data = [] )
    {
        $doctrine = \Config\Services::doctrine();

        if( !isset( $data[ $socialNetwork ] ) ) { return false; }

        $configSocialConnection = config( \ucfirst( $data[ $socialNetwork ] ) . '\\Connection' );
        
        foreach( \get_object_vars( $configSocialConnection ) as $key => $val )
        {
            if( !isset( $value->{ $key } ) )
            {
                return false;
            }

            if( empty( $value->{ $key } ) ){ return false; }

            $configSocialConnection->{ $key } = $value->{ $key };
        }

        return true;
    }**_


Without the change that I propose it does not work

File: tests/system/Validation/ValidationTest.php (hunk 4 above)
Comment (daycry) [reply tests/system/Validation/ValidationTest.php:648]: image

Comment (daycry) [reply tests/system/Validation/ValidationTest.php:648]: I convert to array because the first param must be an array or null

Comment (daycry) [reply tests/system/Validation/ValidationTest.php:648]: I try to simulate an api rest petition.

image

File: tests/system/Validation/ValidationTest.php (hunk 3 above)
Comment (paulbalandan) [reply tests/system/Validation/ValidationTest.php:638]: Hmm. Fair enough. Thanks for the explanation.

**Comment by daycry** [comment]
If you try to validate an object and it fails, the application fails
because it only accepts strings and arrays, it does not do a
json_encode if it is an object, because when you work with rest, you
may have to validate objects.

**Review by paulbalandan — APPROVED** [review]

**Pull request merged**